### PR TITLE
Expanding on roles and decision making

### PR DESCRIPTION
**Pull request by hdgarrood**
See also #1. This PR mostly aims to document what the current roles and
processes for decision making and getting things merged are, so that
they will be clearer to the wider community. That is, most of what I
have written here is intended to describe the current model (rather than
prescribe a new model).

Unlike in #1, I haven't described a "Language Supporter" role: in order
to keep this PR small, I would prefer to focus just on people who have
commit access to the compiler and/or core libraries for now.

There is also a part of this PR which does (sort of) represent a change
in the governance model, however. I would also like to start giving
commit access (i.e. "core collaborator" status) to some people on the
compiler repository. Up until now, we haven't done this at all; everyone
who has commit access on the compiler repository is currently a "core
owner".

I would like to add more core collaborators because we are a little
overstretched: we are not recruiting new maintainers at a very high
rate, and I think there is a bit too much going on for us (i.e. the
current core team) to be able to comfortably keep up with right now.

I think the main barrier to offering more people commit access on the
compiler repo right now is that it is not entirely clear what would be
expected of them. With the changes in this PR, I hope to make
expectations clear, so that we can start adding people as core
collaborators for the compiler repository, and so that these people will
feel empowered to help out without worrying about overstepping
expectations.

Finally, there are also some changes to make the formatting of this
document more consistent, and some minor clarifications (e.g. in the
Purpose section).

### CODE REVIEW

**Comment by hdgarrood** [comment]
I should add that my views on this topic have been shaped a lot by discussions such as https://discourse.purescript.org/t/rfc-documenting-purescripts-governance-structures-policies-processes/1110/ and I would like to thank to the people who were involved for helping me to realise that we do have a project health issue and for suggesting how we can address it, particularly @f-f and @joneshf.

**Review by garyb — APPROVED** [review]
Everything here sounds good to me

File: README.md
Comment (f-f) [inline README.md]: This sounds good to me, but I'd like to specify a timeout for engaging in the discussion.
The reason for this wish is that otherwise it would be ambiguous when a decision is actually effective if not everyone is involved in the discussion, and the risk is for things to get stuck in the "approval limbo"
It doesn't matter what the timeout is exactly (1 week? 2 weeks? 1 month?) because it can be changed if it doesn't fit - it just matters that it's there and it's not unreasonably long (e.g. 1 year would not work)

Two additional small notes:
1. we might want to have something like a set of labels that we attach to PRs and issues to signal the following things:
    - "needs approval from the core team"
    - "voting started"
    - "approved"
    - "rejected"
  
    If this sounds sensible, then we should specify them here so that people involved in such discussions can always refer to this document for reference on how things work.
2. there is a meta aspect involved in all of this, e.g. what's the time frame for this PR to be considered "approved" if not all the owners participate in the discussion? (this is a rhetoric question - the answer is "we approve this one with the implicit rules we had so far")
  However, once this PR is merged then we'll have a documented way to handle approvals/merging/etc. I feel there is a small ambiguity about "is this document included in the rules?" - intuitively it is, but I'd like to see it specified in the "list of decisions that need approval" above

**Review by f-f — APPROVED** [review]
@hdgarrood thank you for this! Very happy to see things moving here :slightly_smiling_face:

I left a comment asking for clarifications, but otherwise this looks good to me :+1:

File: README.md
Comment (hdgarrood) [reply README.md]: I would like to clarify the workflow for issues and PRs, although I think labels might be redundant. Just for the sake of discussion, let's assume that all issues are either feature requests or bug reports, and that people agree on how to categorise them. Then there are three different types of tickets: bug reports, feature requests, and PRs. For bug reports, I don't think we need consider them to be something which need to be voted on, they just need to be validated (or otherwise closed). For feature requests, we've had a vague policy for a while (albeit not always applied consistently and not written down anywhere) that feature requests should go to Discourse first, so that the issue tracker is kept for things that a) represent dev work on the compiler, and b) we are reasonably sure we want to work on. We now have https://github.com/purescript/purescript/issues/3832 for addressing this. For PRs, I think the process is covered already by code review approvals?

Comment (hdgarrood) [reply README.md]: Including changes to this document as things which the "making decisions" process applies to sounds good to me.

Comment (f-f) [reply README.md]: How about the timeout for owners to chime in on things? (which was my main point, sorry if it got lost in the other minor comments)

Comment (natefaubion) [reply README.md]: For PRs, I think we generally wait a few days max after an approval for another owner to voice feedback. I would say 3 days. But we often don't wait for very small changes with an obvious result.

Comment (hdgarrood) [reply README.md]: Yeah, this is more difficult to answer. In general the approach I've taken is to use my judgement - the length of time I'd wait before proceeding with something depends on the potential for decisions to have larger, longer-term consequences. If it's a really big decision I might give it months. I would like people to have time to properly consider proposals and write up any objections properly without feeling rushed, because in my experience it can take a little while to do that.

**Comment by JordanMartinez** [comment]
Why not explicitly say “we'll wait for x days/months for responses for this
issue and will continue if no response received during then. Core owners
can extend this timeline.” ?

>

File: README.md
Comment (hdgarrood) [reply README.md]: Is "approval limbo" a big enough problem that we need to explicitly address it here? I'm not sure it is, but I'm open to being convinced.

Comment (f-f) [reply README.md]: On the other hand I'd say 3 days is very little time for a change that might heavily affect the language/documentation (e.g. what if someone is on vacation for weeks?).
It might be worthwile having different tiers, e.g. 3 days for lightweight decisions and some weeks for big ones, so that people have time to respond

File: README.md
Comment (joneshf) [inline README.md]: Core Collaborators haven't been introduced yet. Maybe we can talk about role changes in a separate section.
```suggestion
```

File: README.md
Comment (joneshf) [inline README.md:85]: After all of the roles have been introduced, we can talk about changing between them.

```suggestion

### Role Changes

*   Core owners may choose to step down and leave the organization at any time.
*   Core collaborators may be promoted to owners at the discretion of the rest of the owners.

```

Comment (joneshf) [reply README.md:85]: Also, there's a way for a Core collaborator to become a Core owner, but no way for a Core owner to become a Core collaborator. Is that intentional, or an oversight?

File: README.md
Comment (joneshf) [inline README.md:100]: Re: [meta issue about this PR](https://github.com/purescript/governance/pull/3#discussion_r409556464)

Correct me if I'm wrong, but I believe @f-f is looking for something along these lines:

```suggestion
*   Promoting collaborators to owners
*   Changes to governance
```

File: README.md
Comment (joneshf) [reply README.md]: I would say that we do need to address approval limbo. I wasn't aware of any times to wait (three days or otherwise), and I've been making more PRs lately. By address, I mean we should at least say what the times are explicitly in this document, not that we need to institute a new process. I think this is @f-f's point.

Labels would also help in mitigating approval limbo. If we label every PR:
* "merge three days after approval"
* "merge two weeks after approval"
* ...

It ought to make this stuff a lot clearer to everyone involved.

Comment (joneshf) [reply README.md]: I also think labeling issues provides value. Right now, it's really hard to know whether an issue was closed because it won't be accepted, isn't a bug, priorities changed, was actually a question, etc. You have to open the issue and read through all the comments to find where the decision was made. A label provides a high-level summary of the decision.

File: README.md
Comment (joneshf) [inline README.md:108]: I'm making a note here, but don't think it needs to be addressed in this PR. We can if we want, but it's not blocking.

We seem to only be talking about the sufficient conditions, but not the necessary conditions. In addition to saying that not everyone has to participate in every discussion, we should state which discussions require a minimum number to participate.

It's been implicit for a long time that at least one other person would be involved in every decision, but never explicit that a decision might need only one person to be decided. There are surely decisions that require a quorum. And there are surely decisions that require a plenum.

Comment (joneshf) [reply README.md:108]: An opinion I have on this is that it would be more straight-forward require all core owners to participate in every discussion. If someone has no opinion on the discussion, they can explicitly approve it because not participating in the discussion is implicit approval (assuming they're aware of the discussion).

I'm not trying to push this opinion, primarily because I highly doubt any of the other core owners agree with it.

**Review by joneshf — COMMENTED** [review]
Thanks for doing this!

File: README.md
Comment (natefaubion) [reply README.md:108]: More than anything I think it's just a bar that won't ever be met. If we predicate being a core owner on absolute participation, then we won't have any core owners. I would love to be able to participate meaningfully in every discussion, but it just can't be done.

Comment (natefaubion) [reply README.md:108]: For this I think it would be helpful to breakdown changes in terms of:
* Maintenance (internal compiler changes) 
* Compiler features (public compiler additions)
* Language additions (non-breaking additions to the surface language and semantics, but may affect tooling)
* Language changes (breaking changes to the surface language and semantics, affects the whole ecosystem)

Comment (joneshf) [reply README.md:108]: Yeah, that would definitely make things easier.

File: README.md
Comment (hdgarrood) [reply README.md:85]: That's an oversight. I think it makes sense for core owners to be able to step down as owners and instead become core collaborators if they wish.

File: README.md
Comment (hdgarrood) [reply README.md:100]: Yes, I just hadn't gotten around to this, thanks!

**Comment by hdgarrood** [comment]
Do people mind if I merge this now and create new issues for the unresolved questions? As far as I can tell, those are:

- Clarify what counts as sufficient time to move forward with a decision
- Clarify what the minimum number of core owner approvals is for moving forward with a decision
- Consider defining milestones or labels in this document for use in GH issue trackers 

Is that everything?

File: README.md
Comment (joneshf) [reply README.md]:
```suggestion
be difficult to reverse later. The current minimum time to wait is three days.
```

**Review by joneshf — CHANGES_REQUESTED** [review]
> * Clarify what counts as sufficient time to move forward with a decision

As mentioned above. I'd like if we addressed this point before merging. Only looking for the time (three days) to be written down, nothing more. I've made a suggestion to help it along.

I'm not particularly fussed on the other points. Feel free to merge and address them later.

**Comment by hdgarrood** [comment]
There are definitely cases where things can be merged immediately, though. This would essentially say that all PRs should wait at least 3 days before merging, right? I'm pretty confident that's not what we want. For instance, just from the last couple of weeks, there was https://github.com/purescript/purescript/pull/3856, https://github.com/purescript/purescript/pull/3853, https://github.com/purescript/purescript/pull/3850, all of which were merged in what I think was an appropriate amount of time.

**Comment by f-f** [comment]
@hdgarrood the original intent about specifying a "timeout" in my review was to put an upper bound, not a lower bound.
I.e. I would like to see a "time limit after which something is automatically approved if not enough voters chime in" - which is to avoid things being stuck.
I'd say it's important to specify _something_ here (even if it's maybe too long, like "1 month"), so that future contributions are subject to this process and timeout.
I agree with you that I don't see the need for a lower bound right away, since small things do not need to wait for much. (though I'd still say we could open an issue to have a discussion there)

**Comment by joneshf** [comment]
Sorry, I misunderstood what the situation was and wrote minimum. I'm not trying to start a different discussion, just messed up.

> I'd say it's important to specify _something_ here (even if it's maybe too long, like "1 month"), so that future contributions are subject to this process and timeout.

Agreed. This is the only thing I'm concerned about. I only made a suggestion so it wasn't just me blocking the merge without trying to move the merge forward. I'd be satisfied with any specific time as long as it's written down.

**Comment by hdgarrood** [comment]
Ok then, I'm happy to say that if something was approved by one or more core owners a month ago and no other core owners have objections, then we can always consider it approved (although we will not expect it to take that long in most cases).

**Comment by hdgarrood** [comment]
Ok, I just pushed a new commit. How's this?

File: README.md
Comment (JordanMartinez) [inline README.md]: Minor typo here: "or choose leave to leave" rather than "or choose to leave"

**Review by joneshf — APPROVED** [review]
I'm good with this. Thanks for making the changes!

File: README.md
Comment (hdgarrood) [reply README.md]: Oops, thanks.

**Pull request merged**
